crypto/dh/dh_gen.c

```diff
@@ -219,6 +220,9 @@ static int dh_builtin_genparams(DH *ret, int prime_len, int generator,
         goto err;
     if (!BN_set_word(ret->params.g, g))
         goto err;
+    /* We are using safe prime p, set key length equivalent to RFC 7919 */
+    ret->length = (2 * ossl_ifc_ffc_compute_security_bits(prime_len)
+                   + 24) / 25 * 25;
```

Comment:
I'm struggling with understanding this computation. How did you arrive at this from RFC7919?

Reply:
@paulidale inferred it from the values given in RFC7919.

Reply:
In 1.1.1 we do not set the length which means we will generate full size keys -> seriously slower than necessary.

The values produced by this formula will be OK for arbitrary safe primes even with prime lengths other than the RFC7919 parameters have. I do not see any reason why it would not be. NIST would allow even smaller values (i.e., the 2*ossl_ifc_ffc_compute_security_bits() values without the rounding up).

Reply:
> In 1.1.1 we do not set the length which means we will generate full size keys -> seriously slower than necessary.

I thought the initial trigger for this was a performance regression in 3.0?

Reply:
The regression is there only for the RFC7919 primes. Not for arbitrary safe primes.

Reply:
The calculation doubles the security bits and rounds up to a multiple of 25.
As @t8m notes, NIST's guidelines suggest only doubling the security bits, so we will be erring conservatively.